### PR TITLE
mgr/dashboard: Fix mypy issues and enable it by default

### PR DESCRIPTION
**Pull request by votdev**
Fix existing mypy issues in the Dashboard Python code and enable mypy by default. This should help to identify Python issues more earlier.

Fixes: https://tracker.ceph.com/issues/44269

Here are some conclusions i made after fixing mypy issues in the Dashboard code:
* Don't reuse a variable if the type changes.
* Use type checking on every new created method. This will help other developers to understand the code. It's really tedious to find out what a method or REST API endpoint returns.
* On the long run we should update all Python code with type checking information.

Signed-off-by: Volker Theile <vtheile@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by votdev** [comment]
@sebastian-philipp ANy idea how to workaround the `dashboard/plugins/plugin.py:25: error: INTERNAL ERROR -- Please try using mypy master on Github:`error? Otherwise this PR can't be merged or we need to remove the `tox.ini`changes. Second issue is that i don't know if there are other issues after the error that are not checked right now.

File: src/pybind/mgr/dashboard/cherrypy_backports.py [hunk 1]
```diff
@@ -174,7 +174,7 @@ def uuid4(self):
                 It's evaluated lazily on render.
                 """
                 try:
-                    self._uuid4
+                    self._uuid4  # type: ignore
```
Comment (votdev) [inline src/pybind/mgr/dashboard/cherrypy_backports.py]: Don't understand why `# type: uuid.UUID` does not make mypy happy.

File: src/pybind/mgr/dashboard/plugins/pluggy.py [hunk 2]
```diff
@@ -75,7 +75,7 @@ class _HookRelay(object):
     """
     def __init__(self):
         from collections import defaultdict
-        self._registry = defaultdict(list)
+        self._registry = defaultdict(list)  # type: ignore
```
Comment (votdev) [inline src/pybind/mgr/dashboard/plugins/pluggy.py]: A `# type: DefaultDict[List]`does not work. Any ideas why?

**Comment by sebastian-philipp** [comment]
> @sebastian-philipp ANy idea how to workaround the `dashboard/plugins/plugin.py:25: error: INTERNAL ERROR -- Please try using mypy master on Github:`error? Otherwise this PR can't be merged or we need to remove the `tox.ini`changes. Second issue is that i don't know if there are other issues after the error that are not checked right now.

easy, just add a 

```python
# type: ignore
```

as the **first** line in `dashboard/plugins/plugin.py`

File: src/pybind/mgr/dashboard/cherrypy_backports.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/cherrypy_backports.py]: which error do you get?

File: src/pybind/mgr/dashboard/plugins/pluggy.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/plugins/pluggy.py]: what do you get?

File: src/pybind/mgr/dashboard/cherrypy_backports.py (hunk 1 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/cherrypy_backports.py]:
```
dashboard/module.py:30: note: In module imported here:
dashboard/cherrypy_backports.py: note: In member "uuid4" of class "LazyUUID4":
dashboard/cherrypy_backports.py:177: error: Cannot determine type of '_uuid4'
Found 1 error in 1 file (checked 9 source files)
```

File: src/pybind/mgr/dashboard/plugins/pluggy.py (hunk 2 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/plugins/pluggy.py]: Fixed with `# type: DefaultDict[str, list]`.

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
> > @sebastian-philipp ANy idea how to workaround the `dashboard/plugins/plugin.py:25: error: INTERNAL ERROR -- Please try using mypy master on Github:`error? Otherwise this PR can't be merged or we need to remove the `tox.ini`changes. Second issue is that i don't know if there are other issues after the error that are not checked right now.
> 
> easy, just add a
> 
> ```python
> # type: ignore
> ```
> 
> as the **first** line in `dashboard/plugins/plugin.py`

I fixed the issue by using the `no_type_check` decorator for the methods which will raise the mypy bug.

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
@votdev: I can run this PR through QA, but `make check` failed and I would like to wait with the QA run for `make check` to pass, so could you take a look at the failure?

**Comment by votdev** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 3]
```diff
@@ -468,7 +468,7 @@ def get(self):
 
 @UiApiController('/cephfs', Scope.CEPHFS)
 class CephFsUi(CephFS):
-    RESOURCE_ID = 'fs_id'
+    RESOURCE_ID = 'fs_id'  # type: ignore
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard/controllers/cephfs.py]:
```suggestion
    RESOURCE_ID = 'fs_id'  # type: str
```
?

File: src/pybind/mgr/dashboard/controllers/iscsi.py [hunk 4]
```diff
@@ -31,6 +36,7 @@ class IscsiUi(BaseController):
 
     @Endpoint()
     @ReadPermission
+    @no_type_check
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard/controllers/iscsi.py]:
```suggestion
    @no_type_check  # https://github.com/python/mypy/issues/7806
```

File: src/pybind/mgr/dashboard/controllers/nfsganesha.py [hunk 5]
```diff
@@ -111,7 +111,7 @@ def status(self):
 @ApiController('/nfs-ganesha/export', Scope.NFS_GANESHA)
 @ControllerDoc(group="NFS-Ganesha")
 class NFSGaneshaExports(RESTController):
-    RESOURCE_ID = "cluster_id/export_id"
+    RESOURCE_ID = "cluster_id/export_id"  # type: ignore
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard/controllers/nfsganesha.py]:
```suggestion
    RESOURCE_ID = "cluster_id/export_id"  # type: str
```
?

File: src/pybind/mgr/dashboard/plugins/debug.py [hunk 6]
```diff
@@ -28,6 +33,7 @@ class Debug(SP, I.CanCherrypy, I.ConfiguresCherryPy):
         )
     ]
 
+    @no_type_check
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard/plugins/debug.py]:
```suggestion
    @no_type_check  # https://github.com/python/mypy/issues/7806
```

File: src/pybind/mgr/dashboard/controllers/cephfs.py (hunk 3 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/controllers/cephfs.py]: Doesn't work, have tried this already.
```
dashboard/plugins/feature_toggles.py:16: note: In module imported here,
dashboard/module.py:56: note: ... from here:
dashboard/controllers/cephfs.py: note: In class "CephFsUi":
dashboard/controllers/cephfs.py:471: error: Incompatible types in assignment (expression has type "str", base class "RESTController" defined the type as "None")
```

File: src/pybind/mgr/dashboard/controllers/nfsganesha.py (hunk 5 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/controllers/nfsganesha.py]: Doesn't work, see https://github.com/ceph/ceph/pull/33454#discussion_r383337694.

File: src/pybind/mgr/dashboard/controllers/cephfs.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/controllers/cephfs.py]: you'll need to annotate 

https://github.com/ceph/ceph/blob/83132bff7e19b2ddd920900377e070150bbc7dd6/src/pybind/mgr/dashboard/controllers/__init__.py#L736

with 

```python
# type: Optional[str]
```

**Comment by callithea** [comment]
jenkins test dashboard

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
- [pkg build](https://shaman.ceph.com/builds/ceph/wip-laura-testing-31922-33454/) [WIP]

File: src/pybind/mgr/dashboard/controllers/cephfs.py (hunk 3 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/controllers/cephfs.py]: Fixed

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
jenkins test dashboard

**Pull request merged**